apidef/graphql_config_adapter.go

```diff
@@ -0,0 +1,280 @@
+		}
+
+		for _, typeField := range ds.RootFields {
+			planTypeField := plan.TypeField{
+				TypeName:   typeField.Type,
+				FieldNames: typeField.Fields,
+			}
+
+			planDataSource.RootNodes = append(planDataSource.RootNodes, planTypeField)
+		}
+
+		switch ds.Kind {
+		case GraphQLEngineDataSourceKindREST:
+			restConfig := GraphQLEngineDataSourceConfigREST{}
+			err = json.Unmarshal(ds.Config, &restConfig)
+			if err != nil {
+				return nil, err
+			}
+
+			factory := &restDataSource.Factory{}
+			if g.httpClient != nil {
+				factory.Client = httpclient.NewNetHttpClient(g.httpClient)
+			}
+			planDataSource.Factory = factory
+
+			planDataSource.Custom = restDataSource.ConfigJSON(restDataSource.Configuration{
+				Fetch: restDataSource.FetchConfiguration{
+					URL:    restConfig.URL,
+					Method: restConfig.Method,
+					Body:   restConfig.Body,
+					Query:  g.convertURLQueriesToEngineV2Queries(restConfig.Query),
+					Header: g.convertHttpHeadersToEngineV2Headers(restConfig.Headers),
+				},
+			})
+
+		case GraphQLEngineDataSourceKindGraphQL:
+
+			graphqlConfig := GraphQLEngineDataSourceConfigGraphQL{}
+			err = json.Unmarshal(ds.Config, &graphqlConfig)
+			if err != nil {
+				return nil, err
+			}
+
+			factory := &graphqlDataSource.Factory{}
+			if g.httpClient != nil {
+				factory.Client = httpclient.NewNetHttpClient(g.httpClient)
+			}
+			planDataSource.Factory = factory
+
+			planDataSource.Custom = graphqlDataSource.ConfigJson(graphqlDataSource.Configuration{
+				Fetch: graphqlDataSource.FetchConfiguration{
+					URL:    graphqlConfig.URL,
+					Method: graphqlConfig.Method,
+					Header: g.convertHttpHeadersToEngineV2Headers(graphqlConfig.Headers),
+				},
+			})
+		}
+
+		planDataSources = append(planDataSources, planDataSource)
+	}
+
+	err = g.determineChildNodes(&planDataSources)
+	return planDataSources, err
+}
+
+func (g *GraphQLConfigAdapter) SetHttpClient(httpClient *http.Client) {
+	g.httpClient = httpClient
+}
+
+func (g *GraphQLConfigAdapter) convertURLQueriesToEngineV2Queries(apiDefQueries map[string]string) []restDataSource.QueryConfiguration {
+	if len(apiDefQueries) == 0 {
+		return nil
+	}
+
+	var engineV2Queries []restDataSource.QueryConfiguration
+	for apiDefQueryName, apiDefQueryValue := range apiDefQueries {
+		engineV2Query := restDataSource.QueryConfiguration{
+			Name:  apiDefQueryName,
+			Value: apiDefQueryValue,
+		}
+
+		engineV2Queries = append(engineV2Queries, engineV2Query)
+	}
+
+	return engineV2Queries
+}
+
+func (g *GraphQLConfigAdapter) convertHttpHeadersToEngineV2Headers(apiDefHeaders map[string]string) http.Header {
+	if len(apiDefHeaders) == 0 {
+		return nil
+	}
+
+	engineV2Headers := make(http.Header)
+	for apiDefHeaderKey, apiDefHeaderValue := range apiDefHeaders {
+		engineV2Headers[apiDefHeaderKey] = []string{apiDefHeaderValue}
+	}
+
+	return engineV2Headers
+}
+
+func (g *GraphQLConfigAdapter) determineChildNodes(planDataSources *[]plan.DataSourceConfiguration) error {
```

Comment:
why we have to pass `planDataSources` as an slice pointer?

Reply:
slice is already a pointer type so we could modify its elements

https://play.golang.org/p/4HnX020UUxF

did I miss anything?

Reply:
You did miss: ...nothing. Pointer was not necessary here!